ensemble_md/utils/coordinate_swap.py

```diff
@@ -0,0 +1,1493 @@
+
+    # Initial position
+    x0, y0, z0 = point
+    init_arry = [x0, y0, z0, 1]
+
+    # Perform the transofmration
+    final_arry = T_r@Rx_r@Ry_r@Rz@Ry@Rx@T@init_arry
+
+    # Get the corrdinates for the rotated point
+    rotated_point = final_arry[0:3]
+
+    return rotated_point
+
+
+def find_rotation_angle(initial_point, vertex, rotated_point, axis):
+    """
+    Determines the angle of rotation around an arbitrary axis in 3D space.
+
+    Parameters
+    ----------
+    initial_point: numpy.ndarray
+        The initial point coordinates.
+    rotated_point: numpy.ndarray
+        The rotated point coordinates.
+    axis: A numpy array
+        The axis of rotation.
+
+    Returns
+    -------
+    angle: int
+        The angle of rotation in radians.
+    """
+    # Normalize axis and seperate coordinates
+    u, v, w = axis / np.linalg.norm(axis)
+    x0, y0, z0 = initial_point
+    a, b, c = vertex
+
+    # Determine the rotated point with the same bond length as the initial point
+    rotated_point_axis = rotated_point - vertex
+    norm_rot_point_vec = rotated_point_axis / np.linalg.norm(rotated_point_axis)
+    final_bond_length = np.sqrt((x0 - a) ** 2 + (y0 - b) ** 2 + (z0 - c) ** 2)
+    rotated_point_standard = vertex + final_bond_length*norm_rot_point_vec
+    x, y, z = rotated_point_standard  # Change to point with same bond length along this vector
+
+    Q = np.sqrt(v ** 2 + w ** 2)
+    B = ((w / Q) * (y0 - b)) - ((v / Q) * (z0 - c))
+    A = Q * (x0 - a) - u * (((v / Q) * (y0 - b)) + ((w / Q) * (z0 - c)))
+    C = u * (x0 - a) + Q * (((v / Q) * (y0 - b)) + ((w / Q) * (z0 - c)))
+    L = np.arctan2(B, A)
+
+    # Determine the rotational angle to line-up point
+    cos_arg = (x - a - u * C) / (Q * np.sqrt(A ** 2 + B ** 2))
+    if cos_arg > 1 or cos_arg < -1:
+        if abs(cos_arg - 1) < .0001:
+            cos_arg = 1
+        elif abs(cos_arg + 1) < .0001:
+            cos_arg = -1
+        else:
+            raise Exception(f'Invalid arccos argument: {(x - a - u * C) / (Q * np.sqrt(A ** 2 + B ** 2))}')
+    arc_cos = np.arccos(cos_arg)
+    angle = L + arc_cos
+
+    # Test point
+    new_pt = rotate_point_around_axis(initial_point, vertex, axis / np.linalg.norm(axis), angle)
+    if abs(new_pt[1] - y) < 0.0001:
+        return angle
+    else:
+        arc_cos = 2*np.pi - arc_cos
+        angle = L + arc_cos
+        if angle > 2*np.pi:
+            angle = angle - 2*np.pi
+
+    return angle
+
+
+def compute_angle(coords):
+    """
+    Computes the angle between two vectors.
+
+    Parameters
+    ----------
+    coords : numpy.ndarray
+        Four points which define two vectors to compute the angle between.
+
+    Returns
+    -------
+    angle : int
+        Angle in radians between the two points.
+    """
+    vec1 = coords[0] - coords[1]
+    vec2 = coords[2] - coords[1]
+
+    angle = np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))

```

Comment:
It does not seem to me that arccos is taken within the function.

Reply:
You are right. I did miss an arccos here. It's in my separate function I was using to run all the production simulations so I must have just missed updating the version here at some point.